setup.sh

```diff
@@ -19,45 +17,71 @@
 # update.
 #
 
-setopt err_exit
+set -ueo pipefail
 
-local projects
-typeset -a projects
-projects=(dmd druntime phobos dlang.org tools installer)
+declare -a projects
+projects=(dmd druntime phobos dlang.org tools installer dub)
 # Working directory
-local wd=$(pwd)
+wd=$(pwd)
+# github username
+githubUser="dlang"
 # Configuration
-local makecmd=make
-local parallel=8
-local model=64
+makecmd="make"
+parallel=8
+model=64
+build="release"
 # List of projects to install vs. update. Their disjoint union is
 # $projects.
-local toInstall toUpdate
-typeset -a toInstall toUpdate
+declare -a toInstall toUpdate
+toInstall=()
+toUpdate=()
 # Mess to go here
-local tempdir=$(mktemp -d /tmp/dmd-update.XXX)
+tempdir=$(mktemp -d /tmp/dmd-update.XXX)
+
+function cleanup() {
+    rm -rf "$tempdir";
+}
+trap cleanup EXIT
+
+function help() {
+    echo "./setup.sh
+Clones and builds dmd, druntime, phobos, dlang.org, tools, installer and dub.
+
+Additional usage
+
+  install       replace current dmd binary with the freshly dmd
+
+Options
+
+  --user=USER   set a custom GitHub user name (requires the repos to be forked)
+  --tag=TAG     select a specific tag to clone"
+}
 
 #
 # Take care of the command line arguments
 #
 function handleCmdLine() {
-    local arg
-    for arg in $*; do
-        case $arg in
-    	    (--tag=*)
-            tag="`echo $arg | sed 's/[-a-zA-Z0-9]*=//'`"
+    for arg in "$@"; do
+        case "$arg" in
+    	    --tag=*)
+    	    tag="${arg//[-a-zA-Z0-9]*=//}"
             ;;
-            (install)
+    	    --user=*)
+    	    githubUser="${arg//[-a-zA-Z0-9]*=//}"
+            ;;
+            install)
             install="yes"
             ;;
-            (*)
+            *)
             echo "Error: $arg not recognized." >&2
+            echo
```

Comment:
Why one echo to stderr, and one to stdout?

`printf` is generally considered better suitable than `echo` in shell scripts.